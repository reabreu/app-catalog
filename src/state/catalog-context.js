import React, { useReducer } from "react";

const CatalogStateContext = React.createContext();
const CatalogDispatchContext = React.createContext();

export const ACTIONS = Object.freeze({
  ADD: "ADD",
  ADD_BY_ID: "ADD_BY_ID",
  SETSTATUS: "SETSTATUS",
});

export const STATUS = Object.freeze({
  IDLE: "IDLE",
  FETCHING: "FETCHING",
  ERROR: "ERROR",
});

// Reducer
function catalogReducer(state, action) {
  switch (action.type) {
    case ACTIONS.ADD: {
      const newApps = {};

      action.payload.forEach((element) => {
        newApps[element.id] = element;
      });

      return { ...state, apps: newApps };
    }
    case ACTIONS.SETSTATUS: {
      return { ...state, status: action.payload };
    }
    case ACTIONS.ADD_BY_ID: {
      return {
        ...state,
        apps: {
          ...state.apps,
          [action.payload.id]: action.payload.data,
        },
      };
    }
    default: {
      throw new Error(`Unhandled action type: ${action.type}`);
    }
  }
}

// Async actions
async function fetchAllApps(dispatch) {
  dispatch({ type: ACTIONS.SETSTATUS, payload: STATUS.FETCHING });

  try {
    const request = await fetch("/apps");
    const res = await request.json();

    dispatch({ type: ACTIONS.SETSTATUS, payload: STATUS.IDLE });
    dispatch({ type: ACTIONS.ADD, payload: res });
  } catch (error) {
    console.log("error", error);
    dispatch({ type: ACTIONS.SETSTATUS, payload: STATUS.ERROR });
  }
}

async function fetchAppById(dispatch, id) {
  dispatch({ type: ACTIONS.SETSTATUS, payload: STATUS.FETCHING });

  try {
    const request = await fetch(`/apps/${id}`);
    const res = await request.json();

    dispatch({ type: ACTIONS.SETSTATUS, payload: STATUS.IDLE });
    dispatch({
      type: ACTIONS.ADD_BY_ID,
      payload: {
        id,
        data: res,
      },
    });
  } catch (error) {
    console.log("error", error);
    dispatch({ type: ACTIONS.SETSTATUS, payload: STATUS.ERROR });
  }
}

const initialState = {
  apps: {},
  status: STATUS.IDLE,
};

function CatalogProvider({ children }) {
  const [state, dispatch] = useReducer(catalogReducer, initialState);

  return (
    <CatalogStateContext.Provider value={state}>
      <CatalogDispatchContext.Provider value={dispatch}>
        {children}
      </CatalogDispatchContext.Provider>
    </CatalogStateContext.Provider>
  );
}

function useCatalogState() {
  const context = React.useContext(CatalogStateContext);
  if (context === undefined) {
    throw new Error("useCountState must be used within a CatalogProvider");
  }
  return context;
}

function useCatalogDispatch() {
  const context = React.useContext(CatalogDispatchContext);
  if (context === undefined) {
    throw new Error("useCountDispatch must be used within a CatalogProvider");
  }
  return context;
}

function useCatalogContext() {
  return [useCatalogState(), useCatalogDispatch()];
}

export {
  CatalogProvider,
  useCatalogContext,
  useCatalogState,
  useCatalogDispatch,
  fetchAllApps,
  fetchAppById,
};
