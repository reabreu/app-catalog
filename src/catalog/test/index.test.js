import React from "react";
import { render } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";
import Catalog from "../index";
import { CatalogProvider } from "../../state/catalog-context";

describe("Catalog spec", () => {
  it("Renders app list on request sucess", async () => {
    const { asFragment, findAllByRole } = render(
      <Router>
        <CatalogProvider>
          <Catalog />
        </CatalogProvider>
      </Router>
    );

    //Rendered list
    expect(await findAllByRole("listitem")).toHaveLength(2);
    expect(asFragment()).toMatchSnapshot();
  });
});
