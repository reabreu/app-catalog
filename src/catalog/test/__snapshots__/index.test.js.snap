// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Catalog spec Renders app list on request sucess 1`] = `
<DocumentFragment>
  <h1
    class="sc-AxjAm jWlVxK"
  >
    App Catalog
  </h1>
  <input
    class="sc-AxmLO fRJPNF"
    placeholder="Search"
    value=""
  />
  <ul
    class="sc-AxirZ essigJ"
  >
    <li
      class="sc-AxiKw cfuaTr"
    >
      <a
        class="sc-Axmtr cIJKbp"
        href="/app/asdm123312asdnboqwedsa"
      >
        <div
          class="sc-AxhCb kqkZWI"
        >
          <img
            alt="prometheus-operator-app-chart"
            class="sc-AxhUy kfGxSb"
            src="https://raw.githubusercontent.com/prometheus/prometheus.github.io/master/assets/prometheus_logo-cb55bb5c346.png"
          />
          <div>
            <p
              class="sc-AxgMl kdnRNu"
            >
              prometheus-operator-app-chart
            </p>
            <p
              class="sc-AxheI gyIMnE"
            >
              0.2.0
            </p>
          </div>
        </div>
      </a>
    </li>
    <li
      class="sc-AxiKw cfuaTr"
    >
      <a
        class="sc-Axmtr cIJKbp"
        href="/app/dsjklnsankj12831209asmdo"
      >
        <div
          class="sc-AxhCb kqkZWI"
        >
          <img
            alt="nginx-ingress-controller-app"
            class="sc-AxhUy kfGxSb"
            src="https://upload.wikimedia.org/wikipedia/commons/thumb/c/c5/Nginx_logo.svg/500px-Nginx_logo.svg.png"
          />
          <div>
            <p
              class="sc-AxgMl kdnRNu"
            >
              nginx-ingress-controller-app
            </p>
            <p
              class="sc-AxheI gyIMnE"
            >
              1.6.9
            </p>
          </div>
        </div>
      </a>
    </li>
  </ul>
</DocumentFragment>
`;
